master/docs/manual/cfg-www.rst

```diff
@@ -266,6 +266,28 @@ The available classes are described here:
             'auth': util.UserPasswordAuth({"homer": "doh!"}),
         }
 
+.. py:class:: buildbot.www.auth.CallableAuth(callable)
+
+    :param callable: callable taking user and password as arguments and returning bool value: ``True`` for the accepted login
+                     and ``False`` for the failed login.
+
+    This authentication class allows you to use any callable that gets username and password as arguments and check if the user
+    can login. You may use it e.g. to check the credentials against external database or file.
+
+    ::
+    
+        def check_credentials(user, password):
+            if user == 'snow' and password == 'white':
+                return True
+            else:
+                return False
+
+        from buildbot.plugins import util
+        c['www'] = {
+            # ...
+            'auth': util.CallableAuth(check_credentials),
+        }
```

Comment:
I usually try to ban callables as much as I can in new Buildbot APIs.

I would prefer an subclassing helper.

```
     class MyAuth(util.CustomAuth):
           def check_credentials(user, password):
                     [...]

     c['www']['auth'] = MyAuth()
```

Implementation could look like this:

```
@implementer(ICredentialsChecker)
class CustomAuth(TwistedICredAuthBase):
    credentialInterfaces = IUsernamePassword
    def requestAvatarId(self, cred):
        if self.callable(cred.username, cred.password):
            return defer.succeed(cred.username)
        return defer.fail(UnauthorizedLogin())

    def __init__(self, **kwargs):
        TwistedICredAuthBase.__init__(
            self,
            [BasicCredentialFactory(b"buildbot")],
            [self],
            **kwargs)
```